doc/designs/ech-api.md

```diff
@@ -0,0 +1,684 @@
+value and the related "singleton" ECHConfigList structure.
+`OSSL_ECHSTORE_make_pemech()` allows the caller to produce a "PEM" data
+structure (conforming to the [PEMECH
+specification](https://datatracker.ietf.org/doc/draft-farrell-tls-pemesni/))
+from the `OSSL_ECHSTORE` entry identified by the `index`. (An `index` of
+`OSSL_ECHSTORE_LAST` will select the last entry.)
+These two APIs will typically be used via the `openssl ech` command line tool.
+
+`OSSL_ECHSTORE_set1_echconfig()` will typically be used by a client to ingest
+the "ech=" SvcParamKey value found in an SVCB or HTTPS RR retrieved from the
+DNS. The resulting set of ECHConfig values can then be associated with an
+`SSL_CTX` or `SSL` structure for TLS connections.
+
+Generally, clients will deal with "singleton" ECHConfigList values, but it is
+also possible (in multi-CDN or multi-algorithm cases), that a client may need
+more fine-grained control of which ECHConfig from a set to use for a particular
+TLS connection. Clients that only support a subset of algorithms can
+automatically make such decisions, however, a client faced with a set of HTTPS
+RR values might (in theory) need to match (in particular) the server IP address
+for the connection to the ECHConfig value via the `public_name` field within
+the ECHConfig value. To enable this selection, the `OSSL_ECHSTORE_get_info()`
+API presents the client with the information enabling such selection, and the
+`OSSL_ECHSTORE_downselect()` API gives the client a way to select one
+particular ECHConfig value from the set stored (discarding the rest).
+
+`OSSL_ECHSTORE_set1_key_and_list()` and `OSSL_ECHSTORE_set1_pemech()` can be
+used to load a private key value and associated "singleton" ECHConfigList.
+Those can be used (by servers) to enable ECH for an `SSL_CTX` or `SSL`
+connection. In addition to loading those values, the application can also
+indicate via `for_retry` which ECHConfig value(s) are to be included in the
+`retry_configs` fallback scheme defined by the ECH protocol.
+
+`OSSL_ECHSTORE_num_keys()` allows a server to see how many usable ECH private
+keys are currently in the store, and `OSSL_ECHSTORE_flush_keys()` allows a
+server to flush keys that are older than `age` seconds.
+
+The APIs the clients and servers can use to associate an `OSSL_ECHSTORE`
+with an `SSL_CTX` or `SSL` structure:
+
+```c
+int SSL_CTX_set_echstore(SSL_CTX *ctx, OSSL_ECHSTORE *es);
+int SSL_set_echstore(SSL *s, OSSL_ECHSTORE *es);
+```
+
+ECH will be enabled for the relevant `SSL_CTX` or `SSL` connection
+when these functions succeed. Any previously associated `OSSL_ECHSTORE`
+will be `OSSL_ECHSTORE_free()`ed.
+
+To access the `OSSL_ECHSTORE` associated with an `SSL_CTX` or
+`SSL` connection:
+
+```c
+OSSL_ECHSTORE *SSL_CTX_get_echstore(const SSL_CTX *ctx);
+OSSL_ECHSTORE *SSL_get_echstore(const SSL_CTX *ctx);
+```
+The resulting `OSSL_ECHSTORE` can be modified and then re-associated
+with an `SSL_CTX` or `SSL` connection.
```

Comment:
Is this a get0 or a get1? I'm confused by the "re-associated" text. It implies that the act of getting disassociates the returned object from the `SSL_CTX`/`SSL`? I might have expected it to simply "get" the value but leave the object associated with the `SSL_CTX`/`SSL`.

Are `OSSL_ECHSTORE` objects ref counted?

Reply:
There's my get/get0/get1 blindness again:-) I've not implemented the `OSSL_ECHSTORE` stuff but will proceed to do that now we seem to be on a similar-enough page. I guess my mental model is the application can use this to get the `OSSL_ECHSTORE` from a connection, then modify that (e.g. down selecting) and use `SSL_set_echstore` to get the new value used.

WRT ref-counting: Not currently, but it might need to be.